content/guides/weird_characters.adoc

```diff
@@ -0,0 +1,846 @@
+
+* https://github.com/clojure/core.async/blob/master/examples/walkthrough.clj[core.async Code Walkthrough]
+* https://github.com/clojure/core.async/wiki[core.async Wiki]
+* <<xref/../core_async_go#,Go Block Best Practices>>
+
+== `<symbol>?` - Predicate Marker
+
+Putting `?` at the end of a symbol is a *naming convention* common across
+many languages that support special characters in their symbol names. It is
+used to indicate that the thing is a predicate, i.e. that it *poses a question*.
+For example, imagine using an API that delt with buffer manipulation:
+[source,clojure]
+----
+(def my-buffer (buffers/create-buffer [1 2 3]))
+(buffers/empty my-buffer)
+----
+At a glance, how would you know if the function `empty` in this case,
+* Returned `true` if the passed in buffer was empty, or,
+* Cleared the buffer
+While the author could have renamed `empty` to `is-empty`, the richness of
+symbol naming in Clojure allows us to express intent more symbolically.
+[source,clojure]
+----
+(def my-buffer (buffers/create-buffer [1 2 3]))
+(buffers/empty? my-buffer)
+false
+----
+This is simply a recommended *convension*, not a *requirement*.
+
+* https://github.com/bbatsov/clojure-style-guide#naming[Clojure Style Guide]
+
+== `<symbol>!` - Unsafe Operations
+
+https://github.com/bbatsov/clojure-style-guide#changing-state-fns-with-exclamation-mark[The Clojure style guide has this to say]:
+
+====
+The names of functions/macros that are not safe in STM transactions
+should end with an exclamation mark (e.g `reset!`).
+====
+You'll most commonly see this appended to function names whose purpose
+is to mutate state, e.g. connecting to a data store, updating an atom or
+closing a file stream
+[source,clojure]
+----
+user=> (def my-stateful-thing (atom 0))
+#'user/my-stateful-thing
+user=> (swap! my-stateful-thing inc)
+1
+user=> @my-stateful-thing
+1
+----
+
+This is simply a recommended *convention* and not a *requirement*
+
+Note that the exclamation mark is often pronounced as bang.
+
+* https://github.com/bbatsov/clojure-style-guide#naming[Clojure Style Guide]
+
+== `_` - Irrelevant var
+
+When you see the underscore character used as function arguments or similar,
+it is a common naming convention for vars or arguments you are not interested in using.
+That is you don't intend to use them, so you aren't really interested in
+thinking of a useful name for them.
+
+This is an example using the `add-watch` function that can be used to add
+callback style behaviour when atoms change value. Imagine, given an atom, we
+want to print the new value every time it changes
+[source,clojure]
+----
+(def value (atom 0))
+
+(add-watch value nil (fn [_ _ _ new-value]
+                       (println new-value))
+
+(reset! value 6)
+; prints 6
+(reset! value 9)
+; prints 9
+----
+`add-watch` takes four arguments, but in our case we only really care about the
+last argument - the new value of the atom.
+
+== `#?` - Standard Reader conditional
```

Comment:
Remove "standard". Also, since the few sections above this are naming conventions, not reader characters, I'd recommend moving the reader conditionals above the earmuffs section.